libtenzir/src/json_parser.cpp

```diff
@@ -0,0 +1,240 @@
+      break; // if the iterator itself errors, the document structure is
+             // invalid.
+    }
+    auto doc = *doc_it;
+    TENZIR_ASSERT(not doc.current_location().error());
+    const auto source = std::string_view{
+      doc.current_location().value_unsafe(),
+      json_line.data() + json_line.size(),
+    };
+    if (auto err = doc.error()) {
+      diagnostic::warning("{}", error_message(err))
+        .note("line {}", lines_processed_)
+        .note("skipped invalid JSON `{}`", truncate(source))
+        .emit(*dh);
+      ++diags_emitted;
+      break;
+    }
+    auto val = doc.get_value();
+    if (auto err = val.error()) {
+      diagnostic::warning("{}", error_message(err))
+        .note("line {}", lines_processed_)
+        .note("skipped invalid JSON `{}`", truncate(source))
+        .emit(*dh);
+      ++diags_emitted;
+      break;
+    }
+    auto parser = doc_parser{json_line, *dh, lines_processed_};
+    auto success = parser.parse_object(val.value_unsafe(), builder.record());
+    if (not success) {
+      builder.remove_last();
+      ++diags_emitted;
+      break;
+    }
+  }
+  if (objects_parsed == 0 and diags_emitted == 0) {
+    diagnostic::warning("line did not contain a single valid JSON object")
+      .note("line {}", lines_processed_)
+      .note("skipped invalid JSON `{}`", truncate(json_line))
+      .emit(*dh);
+  } else if (objects_parsed > 1) {
+    diagnostic::warning("more than one JSON object in line")
+      .note("line {}", lines_processed_)
+      .note("encountered a total of {} objects", objects_parsed)
+      .emit(*dh);
+  }
+  auto truncated_count = stream.truncated_bytes();
+  if (truncated_count > 0 and objects_parsed) {
+    auto truncated_text = std::string_view{
+      json_line.data() + json_line.size() - truncated_count, truncated_count};
+    diagnostic::warning("skipped remaining invalid JSON bytes")
+      .note("line {}", lines_processed_)
+      .note("{} bytes remained", truncated_count)
+      .note("skipped invalid JSON `{}`", truncate(truncated_text))
+      .emit(*dh);
+  }
+}
+
+auto ndjson_parser::validate_completion() const -> void {
+  // noop, just exists for easy of implementation
+}
+
+auto default_parser::parse(const chunk& json_chunk) -> void {
+  // Whether to retry on a capacity error
+  auto retry_capacity_failure = false;
+  // How many documents passed the simdjson batch_size.
+  // Those documents must be skipped in order to not duplicate events.
+  auto completed_documents = size_t{0};
+  buffer_.append(
+    {reinterpret_cast<const char*>(json_chunk.data()), json_chunk.size()});
+  auto view = buffer_.view();
+  do {
+    retry_capacity_failure = false;
+    auto err
+      = json_parser.iterate_many(view.data(), view.length(), current_batch_size)
+          .get(stream_);
+    if (err) {
+      // For the simdjson 3.1 it seems impossible to have an error
+      // returned here so it is hard to understand if we can recover from
+      // it somehow.
```

Comment:
Should we just require a more recent version?

Reply:
See the reply on PR above.